# Real-world Data Wrangling

In [1]:
!python -m pip install kaggle==1.6.12

  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Created wheel for kaggle: filename=kaggle-1.6.12-py3-none-any.whl size=103053 sha256=42d890377179e5fd2951a66f0044a35902f552c1a6e56d403a48a5554acab035
  Stored in directory: /Users/carlosvaleriano/Library/Caches/pip/wheels/73/0c/f9/fa3b1c521dd10f31a9e8a4a38a6910727089616c603a8ce40a
Successfully built kaggle


In [3]:
%pip install ucimlrepo

  Using cached ucimlrepo-0.0.7-py3-none-any.whl.metadata (5.5 kB)
Note: you may need to restart the kernel to use updated packages.


**Note:** Restart the kernel to use updated package(s).

## 1. Gather data

### **1.1.** Problem Statement
How do fluctuations in the 30-year fixed mortgage rate relate to typical home value trends (ZVHI) across U.S. metropolitan markets over time?

### **1.2.** Gather at least two datasets using two different data gathering methods

List of data gathering methods:

- Download data manually
- Programmatically downloading files
- Gather data by accessing APIs
- Gather and extract data from HTML files using BeautifulSoup
- Extract data from a SQL database

Each dataset must have at least two variables, and have greater than 500 data samples within each dataset.

For each dataset, briefly describe why you picked the dataset and the gathering method (2-3 full sentences), including the names and significance of the variables in the dataset. Show your work (e.g., if using an API to download the data, please include a snippet of your code). 

Load the dataset programmtically into this notebook.

#### **Dataset 1**

Type: CSV File

Method: The data was gathered by downloading the Zillow Home Value Index (ZVHI) CSV file manually from the Zillow Research data page and loading it programmatically using Python's `pandas` library.

Dataset variables:

*   `RegionName`: The name of the region or Metropolitan Statistical Area (MSA) (e.g., "Los Angeles, CA").
*   `StateName`: The two-letter abbreviation of the state where the region is located (e.g., "CA").
*   `RegionID`: A unique numerical identifier assigned to each region by Zillow.
*   `SizeRank`: A rank representation based on the population size of the metropolitan area (where 0 is the most populous).
*  `RegionType`: The classification or scale of the geographic region (for example, country for national-level metrics, or msa for Metropolitan Statistical Area).
*   `2000-01-31` to `2026-05-31` (Monthly Date Columns): The Zillow Home Value Index (ZHVI) value representing the typical home value for that region in a given month.

In [1]:
import pandas as pd

# Load the Zillow Home Value Index dataset programmatically
zillow_df = pd.read_csv('Zillow_HomeValueIndex.csv')

# Display the first few rows of the dataframe to verify it loaded correctly
zillow_df.head()


,RegionID,SizeRank,RegionName,RegionType,StateName,2000-01-31,2000-02-29,2000-03-31,2000-04-30,2000-05-31,...,2025-08-31,2025-09-30,2025-10-31,2025-11-30,2025-12-31,2026-01-31,2026-02-28,2026-03-31,2026-04-30,2026-05-31
0,102001,0,United States,country,NaN,124226.850719,124445.405574,124716.445774,125297.773801,125966.812663,...,366338.323837,366583.636859,367024.612896,367710.944113,368513.889428,369279.951949,370005.007795,370503.068810,370583.649909,370320.231928
1,394913,1,"New York, NY",msa,NY,220241.022643,221176.922725,222121.490928,224035.551610,226017.627731,...,702535.550236,703949.305049,706306.806277,709680.532631,713132.753507,716225.158400,719831.723350,723655.272546,726086.278463,727625.487581
2,753899,2,"Los Angeles, CA",msa,CA,225669.811207,226509.589232,227628.165553,229853.756673,232287.855633,...,957302.987799,958379.278078,960840.171876,964218.923597,968252.041247,970828.794664,972083.425340,971743.362813,970091.338711,968608.414008
3,394463,3,"Chicago, IL",msa,IL,156798.618056,156943.767900,157220.208333,157907.923431,158735.161975,...,341428.373236,342790.550639,344123.864996,345705.325594,347475.565448,349131.015117,350915.816595,352352.255440,353373.098703,353998.242501
4,394514,4,"Dallas, TX",msa,TX,131332.841965,131391.338072,131458.585335,131633.219413,131863.469663,...,373414.212536,372748.386147,372389.688238,372118.383227,371894.357439,371448.263585,370856.387408,369771.000631,368421.163266,366823.210136


#### Dataset 2

Type: CSV File

Method: The data was gathered using the "Programmatic Downloading" method. The CSV file was retrieved directly from the Federal Reserve Economic Data (FRED) public database URL using Python's `pandas.read_csv()` function.

Dataset variables:
*   `observation_date`: The date of the weekly interest rate measurement (formatted as YYYY-MM-DD).
*   `MORTGAGE30US`: The average interest rate for a 30-year fixed-rate mortgage in the United States (expressed as a percentage, e.g., 7.33%).

In [2]:
import pandas as pd

# Define the programmatic download URL for FRED mortgage rate data
fred_url = "https://fred.stlouisfed.org/graph/fredgraph.csv?id=MORTGAGE30US"

# Load the dataset programmatically
mortgage_df = pd.read_csv(fred_url)

# Display the dataset shape and first few rows to verify loading
print(f"Dataset shape: {mortgage_df.shape}")
mortgage_df.head()


Dataset shape: (2894, 2)


,observation_date,MORTGAGE30US
0,1971-04-02,7.33
1,1971-04-09,7.31
2,1971-04-16,7.31
3,1971-04-23,7.31
4,1971-04-30,7.29


## 2. Assess data

Assess the data according to data quality and tidiness metrics using the report below.

List **two** data quality issues and **two** tidiness issues. Assess each data issue visually **and** programmatically, then briefly describe the issue you find.  **Make sure you include justifications for the methods you use for the assessment.**

### Quality Issue 1:

In [3]:
# Display a slice of the dataframe containing missing values (NaNs)
# We look at the first 5 rows with missing values and show metadata + some early date columns
zillow_df[zillow_df.isnull().any(axis=1)].iloc[:5, :12]

,RegionID,SizeRank,RegionName,RegionType,StateName,2000-01-31,2000-02-29,2000-03-31,2000-04-30,2000-05-31,2000-06-30,2000-07-31
0,102001,0,United States,country,NaN,124226.850719,124445.405574,124716.445774,125297.773801,125966.812663,126684.112495,127443.873514
4,394514,4,"Dallas, TX",msa,TX,131332.841965,131391.338072,131458.585335,131633.219413,131863.469663,132096.530458,132340.891513
11,394976,11,"Phoenix, AZ",msa,AZ,144621.010629,144914.489697,145300.990046,146094.712435,146967.490977,147709.454054,148446.804078
33,394705,34,"Indianapolis, IN",msa,IN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
35,395059,36,"San Jose, CA",msa,CA,375013.804973,376946.593470,379858.303986,387743.425193,396010.491678,405482.558846,413637.153041


In [4]:
# Programmatically calculate the total number of missing values in home value columns
total_nulls = zillow_df.iloc[:, 5:].isnull().sum().sum()
total_cells = zillow_df.shape[0] * zillow_df.iloc[:, 5:].shape[1]
pct_missing = (total_nulls / total_cells) * 100

print(f"Total missing index values: {total_nulls} out of {total_cells}")
print(f"Percentage of missing data: {pct_missing:.2f}%")


Total missing index values: 49266 out of 283715
Percentage of missing data: 17.36%


Issue and justification: The Zillow Home Value Index dataset contains a large number of missing values (49,266 NaNs, or 17.36% of the home price cells). This occurs because smaller or newer metropolitan statistical areas (MSAs) do not have complete historical data dating back to 2000.

Visual inspection using a subset filter allows us to see how these NaNs are distributed in the columns. Programmatic inspection counts the exact volume of missing values so we can make an informed decision on whether to drop incomplete regions or fill the gaps.

### Quality Issue 2:

In [ ]:
# Visually inspect the formatting of the date columns
print("Zillow Date Column Header Examples (first 5 dates):")
print(list(zillow_df.columns[5:10]))

print("\nMortgage Date Column Values (first 5):")
print(mortgage_df['observation_date'].head())


Zillow Date Column Header Examples (first 5):
['2000-01-31', '2000-02-29', '2000-03-31', '2000-04-30', '2000-05-31']

Mortgage Date Column Values (first 5):
0    1971-04-02
1    1971-04-09
2    1971-04-16
3    1971-04-23
4    1971-04-30
Name: observation_date, dtype: object


In [6]:
# Programmatically inspect the exact data types of the date variables
print(f"Mortgage 'observation_date' data type: {mortgage_df['observation_date'].dtype}")

Mortgage 'observation_date' data type: object


Issue and justification: The date columns in both datasets are represented as string objects instead of `datetime64`. For the Zillow dataset, the dates are stored as column headers. For the Mortgage dataset, the `observation_date` is a text column. 

Visual inspection shows the dates are structured as text. Programmatic inspection verifies that Pandas recognizes them as strings (`object`). We must convert these to datetime objects to perform chronological sorting, time-series plotting, and merging operations.

### Tidiness Issue 1:

Here are the three essential rules of data tidiness:

1. Each variable forms a column: Each unique variable in the dataset should be in its own column.
2. Each observation forms a row: Each individual data observation should have its own row.
3. Each type of observational unit forms a table: Different types of data (e.g., customers and products) should be in separate tables unless they are directly related.

In [ ]:
# Inspecting the dataframe visually
zillow_df.head()

,RegionID,SizeRank,RegionName,RegionType,StateName,2000-01-31,2000-02-29,2000-03-31,2000-04-30,2000-05-31,...,2025-08-31,2025-09-30,2025-10-31,2025-11-30,2025-12-31,2026-01-31,2026-02-28,2026-03-31,2026-04-30,2026-05-31
0,102001,0,United States,country,NaN,124226.850719,124445.405574,124716.445774,125297.773801,125966.812663,...,366338.323837,366583.636859,367024.612896,367710.944113,368513.889428,369279.951949,370005.007795,370503.068810,370583.649909,370320.231928
1,394913,1,"New York, NY",msa,NY,220241.022643,221176.922725,222121.490928,224035.551610,226017.627731,...,702535.550236,703949.305049,706306.806277,709680.532631,713132.753507,716225.158400,719831.723350,723655.272546,726086.278463,727625.487581
2,753899,2,"Los Angeles, CA",msa,CA,225669.811207,226509.589232,227628.165553,229853.756673,232287.855633,...,957302.987799,958379.278078,960840.171876,964218.923597,968252.041247,970828.794664,972083.425340,971743.362813,970091.338711,968608.414008
3,394463,3,"Chicago, IL",msa,IL,156798.618056,156943.767900,157220.208333,157907.923431,158735.161975,...,341428.373236,342790.550639,344123.864996,345705.325594,347475.565448,349131.015117,350915.816595,352352.255440,353373.098703,353998.242501
4,394514,4,"Dallas, TX",msa,TX,131332.841965,131391.338072,131458.585335,131633.219413,131863.469663,...,373414.212536,372748.386147,372389.688238,372118.383227,371894.357439,371448.263585,370856.387408,369771.000631,368421.163266,366823.210136


Observation: There are hundreds of monthly date columns.

In [7]:
# Inspecting the dataframe programmatically
zillow_df.shape

(895, 322)

Issue and justification: Monthly dates are spread across 300+ columns rather than existing in a single Date column. Column headers should be variable names, not date or data values. Having dates as columns causes issues like not being able to merge on date or plotting time series. 

### Tidiness Issue 2: 

In [8]:
# Inspecting the dataframe visually
mortgage_df.head()
zillow_df['RegionType'].value_counts()

RegionType
msa        894
country      1
Name: count, dtype: int64

In [9]:
# Inspecting the dataframe programmatically
mortgage_df['observation_date']

0       1971-04-02
1       1971-04-09
2       1971-04-16
3       1971-04-23
4       1971-04-30
           ...    
2889    2026-08-13
2890    2026-08-20
2891    2026-08-27
2892    2026-09-03
2893    2026-09-10
Name: observation_date, Length: 2894, dtype: object

Issue and justification: FRED mortgage rates are measured weekly as opposed to Zillow measuring them monthly. Also, Zillow contains national and metropolitan level records in the same table as seen through the 'RegionType' column.

To merge the mortgage rates with home values, the weekly mortgage rates should be resampled to monthly averages. 

## 3. Clean data
Clean the data to solve the 4 issues corresponding to data quality and tidiness found in the assessing step. **Make sure you include justifications for your cleaning decisions.**

After the cleaning for each issue, please use **either** the visually or programatical method to validate the cleaning was succesful.

At this stage, you are also expected to remove variables that are unnecessary for your analysis and combine your datasets. Depending on your datasets, you may choose to perform variable combination and elimination before or after the cleaning stage. Your dataset must have **at least** 4 variables after combining the data.

In [ ]:
# FILL IN - Make copies of the datasets to ensure the raw dataframes 
# are not impacted

### **Quality Issue 1: FILL IN**

In [ ]:
# FILL IN - Apply the cleaning strategy

In [ ]:
# FILL IN - Validate the cleaning was successful

Justification: *FILL IN*

### **Quality Issue 2: FILL IN**

In [ ]:
#FILL IN - Apply the cleaning strategy

In [ ]:
#FILL IN - Validate the cleaning was successful

Justification: *FILL IN*

### **Tidiness Issue 1: FILL IN**

In [ ]:
#FILL IN - Apply the cleaning strategy

In [ ]:
#FILL IN - Validate the cleaning was successful

Justification: *FILL IN*

### **Tidiness Issue 2: FILL IN**

In [ ]:
#FILL IN - Apply the cleaning strategy

In [ ]:
#FILL IN - Validate the cleaning was successful

Justification: *FILL IN*

### **Remove unnecessary variables and combine datasets**

Depending on the datasets, you can also peform the combination before the cleaning steps.

In [ ]:
#FILL IN - Remove unnecessary variables and combine datasets

## 4. Update your data store
Update your local database/data store with the cleaned data, following best practices for storing your cleaned data:

- Must maintain different instances / versions of data (raw and cleaned data)
- Must name the dataset files informatively
- Ensure both the raw and cleaned data is saved to your database/data store

In [ ]:
#FILL IN - saving data

## 5. Answer the research question

### **5.1:** Define and answer the research question 
Going back to the problem statement in step 1, use the cleaned data to answer the question you raised. Produce **at least** two visualizations using the cleaned data and explain how they help you answer the question.

*Research question:* FILL IN from answer to Step 1

In [ ]:
#Visual 1 - FILL IN

*Answer to research question:* FILL IN

In [ ]:
#Visual 2 - FILL IN

*Answer to research question:* FILL IN

### **5.2:** Reflection
In 2-4 sentences, if you had more time to complete the project, what actions would you take? For example, which data quality and structural issues would you look into further, and what research questions would you further explore?

*Answer:* FILL IN